runtime/lua/vim/lsp.lua

```diff
@@ -358,16 +362,47 @@ do
     end
   end
 
-  function changetracking.prepare(bufnr, firstline, lastline, new_lastline, changedtick)
+  function changetracking.prepare(bufnr, firstline, old_lastline, new_lastline, changedtick)
     local incremental_changes = function(client)
-      local cached_buffers = state_by_client[client.id].buffers
-      local curr_lines = nvim_buf_get_lines(bufnr, 0, -1, true)
-      local line_ending = format_line_ending[vim.api.nvim_buf_get_option(0, 'fileformat')]
+      local state = state_by_client[client.id]
+      local prev_lines = state.buffers[bufnr]
+      -- See the comment below for why the curr_lines table is pulled out of buffers_tmp
+      -- instead of making a brand new table.
+      local curr_lines = state.buffers_tmp[bufnr]
+
+      local changed_lines = nvim_buf_get_lines(bufnr, firstline, new_lastline, true)
+      for i = 1, firstline do
+        curr_lines[i] = prev_lines[i]
+      end
+      for i = firstline + 1, new_lastline do
+        curr_lines[i] = changed_lines[i - firstline]
+      end
+      for i = old_lastline + 1, #prev_lines do
+        curr_lines[i - old_lastline + new_lastline] = prev_lines[i]
+      end
+      if tbl_isempty(curr_lines) then
+        -- Can happen when deleting the entire contents of a buffer, see https://github.com/neovim/neovim/issues/16259.
+        curr_lines[1] = ''
+      end
+
+      local line_ending = buf_get_line_ending(bufnr)
       local incremental_change = sync.compute_diff(
-        cached_buffers[bufnr], curr_lines, firstline, lastline, new_lastline, client.offset_encoding or 'utf-16', line_ending or '\n')
-      cached_buffers[bufnr] = curr_lines
+        prev_lines, curr_lines, firstline, old_lastline, new_lastline, client.offset_encoding or 'utf-16', line_ending)
+
+      -- Double-buffering of lines tables is used to reduce the load on the garbage collector.
+      -- At this point the prev_lines table is useless, but its internal storage has already been allocated,
+      -- so let's keep it around for the next didChange event, in which it will become the next
+      -- curr_lines table. Note that setting elements to nil doesn't actually deallocate slots in the
+      -- internal storage - it merely marks them as free, for the GC to deallocate them.
+      for i = #prev_lines, 1, -1 do
```

Comment:
Using a counter is faster than ipairs on PUC Lua 5.1 ([source](https://gitspartv.github.io/LuaJIT-Benchmarks/#test10)), and I imagine this loop would be quite the hot one in files with lots of lines.

Reply:
Not sure if mentioned anywhere, but if you can measure this for the large slowdowns and it provides a meaningful difference, then you can keep. But otherwise, I think it obscures what is happening. I find it unlikely it will provide any meaningful difference (but it's possible I suppose)

Reply:
Alright, let's compare the table clearing loop using ipairs vs the current implementation with a counter.

```diff
diff --git a/runtime/lua/vim/lsp.lua b/runtime/lua/vim/lsp.lua
index 9eaa41e75..5d6bf6eb5 100644
--- a/runtime/lua/vim/lsp.lua
+++ b/runtime/lua/vim/lsp.lua
@@ -389,7 +389,7 @@ do
       -- so let's keep it around for the next didChange event, in which it will become the next
       -- curr_lines table. Note that setting elements to nil doesn't actually deallocate slots in the
       -- internal storage - it merely marks them as free, for the GC to deallocate them.
-      for i = #prev_lines, 1, -1 do
+      for i in ipairs(prev_lines) do
         prev_lines[i] = nil
       end
       state.buffers[bufnr] = curr_lines
```

<details><summary>Benchmarking method</summary>

The benchmark is almost the same as I have described in the commit message, with the exception that I increased the number of runs to 40 and swapped vim-commentary for `call setline(1, getline(1, '$'))` - as I had explained [here](https://github.com/neovim/neovim/pull/16277#issuecomment-968105798), it exhibits the exact same problem, but removes the Vimscript interpreter from the equation, thus I need to wait less for the benchmark to complete.

```vim
set rtp+=~/.config/nvim/plugged/nvim-lspconfig
lua require('lspconfig')['ccls'].setup({})
function! Benchmark(tries) abort
  let results_comment = []
  let results_undo = []
  for i in range(a:tries)
    echo printf('run %d', i+1)
    let begin = reltime()
    call setline(1, getline(1, '$'))
    call add(results_comment, reltimefloat(reltime(begin)))
    let begin = reltime()
    silent! undo
    call add(results_undo, reltimefloat(reltime(begin)))
    redraw
  endfor
  let avg_comment = 0.0
  let avg_undo = 0.0
  for i in range(a:tries)
    echomsg printf('run %3d: setline=%fs undo=%fs', i+1, results_comment[i], results_undo[i])
    let avg_comment += results_comment[i]
    let avg_undo += results_undo[i]
  endfor
  echomsg printf('average: setline=%fs undo=%fs', avg_comment / a:tries, avg_undo / a:tries)
endfunction
command! -bar Benchmark call Benchmark(40)
```

I have performed tests on both Lua5.1 and LuaJIT. I built nvim with LuaJIT as explained in the commit (`make CMAKE_BUILD_TYPE=Release`), and with Lua5.1 like this:

```sh
make CMAKE_BUILD_TYPE=Release CMAKE_EXTRA_FLAGS='-DPREFER_LUA=ON' DEPS_CMAKE_FLAGS='-DUSE_BUNDLED_LUA=ON -DUSE_BUNDLED_LUAJIT=OFF'
```

Then opened the file `src/nvim/lua/executor.c` in the nvim source tree and ran `:Benchmark`.

</details>

Here are the results:

| Lua version | table clearing loop | setline time | undo time |
| ----------- | ------------------- | ------------ | --------- |
| Lua5.1      | with ipairs         | 0.276445s    | 0.270881s |
| Lua5.1      | with a counter      | 0.190557s    | 0.185333s |
| LuaJIT      | with ipairs         | 0.047433s    | 0.042246s |
| LuaJIT      | with a counter      | 0.047436s    | 0.042917s |

Notice that on LuaJIT the difference is next to none, as the microbenchmark I've linked in a comment above suggests.

Reply:
I would just use ipairs then

Reply:
The conclusion we have reached in the Matrix room is that if you are using Lua5.1 builds you are a priori not interested in performance. I have also measured the difference on LuaJIT 2.0.5 (as opposed to the beta version Nvim downloads) to be ~2% faster with a counter, which is negligible IMO. (Then again I am surprised that this loop is such a big deal.)